### PR TITLE
[FIX] Custom Oauth store refresh and id tokens with expiresIn

### LINKED ISSUE #13693: [BUG] Custom Oauth does not register the refreshToken nor respect the accessToken lifetime.
### Description:
- Custom Oauth implementation does include the refreshToken in it's call to registerService(), as seen [here](https://github.com/RocketChat/Rocket.Chat/blob/afb439f27b77e0ff0db9c4087f441518847f5e7b/packages/rocketchat-custom-oauth/server/custom_oauth_server.js#L248)
- RocketChat server seems to be handling the session lifetime on it's own, as opposed to respecting the accessToken lifetime, using the refreshToken to update the accessToken, and forcing the user to re-login in the case that they have both expired. 
### Steps to reproduce:
1. Login with a customOauth provider and you will find that you can remain logged in to the server forever, unless you purposefully logout.  This includes when the client app (in our case android) is closed and reopened on the device. Basically, the call to isAuthenticated() always comes back 'true', unless you have purposefully logged out at some point. 
### Expected behavior:
- I would expect that the server should respect the accessToken lifetime (in our case 2 hours). 
- When a client opens a new session, if the accessToken has expired, the RC server should request a new access token from the Oauth provider, using the refreshToken (in our case the refreshToken has a lifetime of 30 days). 
- The RC server would then receive both a new accessToken and refreshToken in the response, which should be re-registered with the service. 
### Actual behavior:
- Once logged in, a client can remain logged in forever, no matter the state of the accessToken or the refreshToken.

### PR DESCRIPTION
**Pull request by ralfbecker**
Closes #13693

### CODE REVIEW

**Comment by ralfbecker** [comment]
@geekgonecrazy do you might reviewing this

**Comment by geekgonecrazy** [comment]
Nice!  I’m not sure however... if the refresh is ever even used?

**Comment by ralfbecker** [comment]
The ci/circleci test failures have nothing to do with this pull-request:
```
WARNING: The following packages cannot be authenticated!
  google-chrome-stable
E: There were unauthenticated packages and -y was used without --allow-unauthenticated
```

**Comment by ralfbecker** [comment]
@geekgonecrazy You mean if Rocket.Chat/Meteor uses refreshTokens, or that it need to be set in a different way? I'm setting it now, like the Google provider sets it.

Ralf

**Comment by geekgonecrazy** [comment]
No I mean.  Now we have it.. cool!  **but** where or how is the refresh token actually used?

**Comment by ralfbecker** [comment]
The accessToken is registered with it's expiration. If we got an refreshToken, we can use that to get a new accessToken, after it's expired, without "bugging" the user again.
My expectation is that this is done in Meteor's or Rocket.Chat's existing OAuth code. Why else would the other providers register the refreshToken. I have not verified that!

Ralf

**Comment by geekgonecrazy** [comment]
Just a heads up by adding the registerAccessToken as part of this you have overlap with: #14113 opened just a bit before yours.

**Comment by ralfbecker** [comment]
Hmm, #14113 seems to do the same as my intended next pull request:
https://github.com/RocketChat/Rocket.Chat/commit/63f6e52fc8b01449dd2cd392ed7e0a129283ee47

File: app/custom-oauth/server/custom_oauth_server.js [hunk 1]
```diff
@@ -254,9 +254,18 @@ export class CustomOAuth {
 
 			const serviceData = {
 				_OAuthCustom: true,
-				accessToken,
+				accessToken: response.access_token,
+				idToken: response.id_token,
+				expiresAt: (+new Date) + (1000 * parseInt(response.expires_in, 10)),
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]: We may need to add something to actually honor this.  I'm looking through meteors code base and don't see anything that even honors expiresAt :(

Comment (ralfbecker) [reply app/custom-oauth/server/custom_oauth_server.js]: I'll have a look tomorrow, and create a pull request to use expiresAt and refreshToken, if I can't find it's already there.

Btw. can you point me to some docu on how to build a Docker container from my changed sources / git.

Comment (geekgonecrazy) [reply app/custom-oauth/server/custom_oauth_server.js]: do `meteor build --directory /folder/you/want/to/build/in`

Then `cp .docker/Dockerfile /folder/you/want/to/build/in`

finally: go into the folder and `docker build -t theimage:tag .`

Should do the trick

**Comment by ralfbecker** [comment]
I rebased this pull request in my repo on top of @knrt10 one: https://github.com/EGroupware/Rocket.Chat/commit/7aba9cad805d376aaef2c7a7e6901bc742df00ad

I searched through Rocket.Chat codebase for usage of expiresAt or refreshToken: I'm pretty sure they are NOT used at all :(

I think when Rocket.Chat receives an accessToken, eg. through custom OAuth, it's ok to verify it's currently valid by using the identity endpoint with the accessToken as Authorization: Bearer. 
Rocket.Chat then stores, with this pull request, expiresAt and, if we receive one, the refreshToken.

When the next Api request with the same token arrives, you can either:
a) check it again with the identity endpoint or
b) check the stored expiresAt time and only re-check if it's past that time
So for Api access there is (only) less latency to again, refresh token does not help at all.

Other use case is the Rocket.Chat client and I believe that's what https://github.com/RocketChat/Rocket.Chat/issues/13693 is talking about: after the initial OAuth authentication by the user, Rocket.Chat should not assume the user is authenticated for ever, but only for the lifetime of the token. If the token is expired, Rocket.Chat either needs to bug the user again for an other OAuth authentication, or try to automatically get a new accessToken through the refreshToken (from the server-side), which might fail in case the user revoked the refreshToken in the meantime because he eg. lost the device. 

I think that's were Rocket.Chat has a problem: there is no way to cut off access by a lost client-device, in case an external OAuth server is used.

Ralf

**Comment by geekgonecrazy** [comment]
Seems you are at the same conclusion I reached.

I still think this is great to have added.  But yes we may need to follow up with a pull request to tie rocket.chat token life time to oauth life time.  But I have a feeling that will not be an easy task to track down

File: app/custom-oauth/server/custom_oauth_server.js [hunk 2]
```diff
@@ -182,14 +182,24 @@ export class CustomOAuth {
 	registerService() {
 		const self = this;
 		OAuth.registerService(this.name, 2, null, (query) => {
-			const accessToken = self.getAccessToken(query);
 
-			const identity = self.getIdentity(accessToken, this.accessTokenParam);
+			const response = self.getAccessToken(query);
+
+			const identity = self.getIdentity(response.accessToken, this.accessTokenParam);
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
			const identity = self.getIdentity(response.access_token, this.accessTokenParam);
```

Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
			const identity = self.getIdentity(response.access_token, this.accessTokenParam);
```

File: app/custom-oauth/server/custom_oauth_server.js [hunk 3]
```diff
@@ -182,14 +182,24 @@ export class CustomOAuth {
...
 
 			const serviceData = {
 				_OAuthCustom: true,
-				accessToken,
+				accessToken: response.access_token,
+        idToken: response.id_token,
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
				idToken: response.id_token,
```

File: app/custom-oauth/server/custom_oauth_server.js [hunk 4]
```diff
@@ -182,14 +182,24 @@ export class CustomOAuth {
...
+				expiresAt: (+new Date) + (1000 * parseInt(response.expires_in, 10)),
 			};
+      
+      // only set the token in serviceData if it's there. this ensures
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
			// only set the token in serviceData if it's there. this ensures
```

**Comment by geekgonecrazy** [comment]
Sorry some of these suggestions are only needed because of the conflict merging.  😊

**Comment by weand** [comment]
> Seems you are at the same conclusion I reached.
> 
> I still think this is great to have added. But yes we may need to follow up with a pull request to tie rocket.chat token life time to oauth life time. But I have a feeling that will not be an easy task to track down

is there any issue available thats tracks progress for _"pull request to tie rocket.chat token life time to oauth life time"_

**Pull request merged**